monai/apps/detection/utils/anchor_utils.py

```diff
@@ -0,0 +1,408 @@
+
+# BSD 3-Clause License
+
+# Copyright (c) Soumith Chintala 2016,
+# All rights reserved.
+
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+
+"""
+This script is adapted from
+https://github.com/pytorch/vision/blob/release/0.12/torchvision/models/detection/anchor_utils.py
+"""
+
+from typing import List, Sequence, Union
+
+import torch
+from torch import Tensor, nn
+
+from monai.utils import ensure_tuple
+from monai.utils.misc import issequenceiterable
+from monai.utils.module import look_up_option
+
+
+class AnchorGenerator(nn.Module):
+    """
+    This module is modified from torchvision to support both 2D and 3D images.
+
+    Module that generates anchors for a set of feature maps and
+    image sizes.
+
+    The module support computing anchors at multiple sizes and aspect ratios
+    per feature map.
+
+    sizes and aspect_ratios should have the same number of elements, and it should
+    correspond to the number of feature maps.
+
+    sizes[i] and aspect_ratios[i] can have an arbitrary number of elements.
+    For 2D images, anchor width and height w:h = 1:aspect_ratios[i,j]
+    For 3D images, anchor width, height, and depth w:h:d = 1:aspect_ratios[i,j,0]:aspect_ratios[i,j,1]
+
+    AnchorGenerator will output a set of sizes[i] * aspect_ratios[i] anchors
+    per spatial location for feature map i.
+
+    Args:
+        sizes: base size of each anchor.
+            len(sizes) is the number of feature maps, i.e., the number of output levels for
+            the feature pyramid network (FPN).
+            Each elment of ``sizes`` is a Sequence which represents several anchor sizes for each feature map.
+        aspect_ratios: the aspect ratios of anchors. ``len(aspect_ratios) = len(sizes)``.
+            For 2D images, each element of ``aspect_ratios[i]`` is a Sequence of float.
+            For 3D images, each element of ``aspect_ratios[i]`` is a Sequence of 2 value Sequence.
+        indexing: choose from {'xy', 'ij'}, optional
+            Cartesian ('xy') or matrix ('ij', default) indexing of output.
+            Cartesian ('xy') indexing swaps axis 0 and 1, which is the setting inside torchvision.
+            matrix ('ij', default) indexing keeps the original axis not changed.
+            See also indexing in https://pytorch.org/docs/stable/generated/torch.meshgrid.html
```

Comment:
I find this a bit confusion, maybe the documentation could be improved here? The indexing of torch.meshgrid is always performed with 'ij' in the code and the axis are manually swapped. Does this also influence the aspect ratios and scales?

Reply:
Yes, it is confusing. But that's how it works in torchvision. In the unit test, i tested its output and made sure it agrees with torchvision for 2D cases.

Reply:
I added more docstring to explain our class with indexing 'xy' is equivalent to torchvision in 2D cases